Sources/SecureXPC/Client/XPCClient.swift

```diff
@@ -224,41 +227,60 @@ public class XPCClient {
             } else { // Unexpected
                 result = Result.failure(.unknown)
             }
+            self.handleConnectionErrors(event: xpcResponse)
             reply(result)
         })
     }
 
-    private func getConnection() -> xpc_connection_t {
+    private func getConnection() throws -> xpc_connection_t {
         if let existingConnection = self.connection { return existingConnection }
 
-        let newConnection = self.createConnection()
+        let newConnection = try self.createConnection()
         self.connection = newConnection
 
-        xpc_connection_set_event_handler(newConnection, self.handle(connectionEvent:))
+        xpc_connection_set_event_handler(newConnection, self.handleConnectionErrors(event:))
         xpc_connection_resume(newConnection)
 
         return newConnection
     }
 
-    private func handle(connectionEvent: xpc_object_t) {
-        if xpc_equal(connectionEvent, XPC_ERROR_CONNECTION_INVALID) {
-            self.connection = nil
-        } else if xpc_equal(connectionEvent, XPC_ERROR_CONNECTION_INTERRUPTED) {
-            self.connection = nil
-        } else if xpc_equal(connectionEvent, XPC_ERROR_TERMINATION_IMMINENT) {
+    private func handleConnectionErrors(event: xpc_object_t) {
+        if xpc_equal(event, XPC_ERROR_CONNECTION_INVALID) {
+            // Paraphrasing from Apple documentation:
+            //   If the named service provided could not be found in the XPC service namespace. The connection is
+            //   useless and should be disposed of.
+            //
+            // While the underlying connection is useless, this client instance is *not* useless. A scenario we want to
+            // support is:
+            //  - API user creates a client
+            //  - Attempts to send a message to a blessed helper tool
+            //  - `XPCError.connectionInvalid` is thrown
+            //  - Error is handled by installing the helper tool
+            //  - Using the same client instance successfully sends a message to the now installed helper tool
             self.connection = nil
+        } else if xpc_equal(event, XPC_ERROR_CONNECTION_INTERRUPTED) {
+            // Apple documentation:
+            //   Will be delivered to the connection’s event handler if the remote service exited. The connection is
+            //   still live even in this case, and resending a message will cause the service to be launched on-demand.
+            //
+            // While Apple's documentation is technically correct, it's misleading in the case of an anonymous
+            // connection where there is no service. Because there is no service, there is nothing to be relaunched
+            // on-demand. The connection might technically still be alive, but resending a message will *not* work.
+            //
+            // By setting the connection to `nil` when there is no service (indicated by no service name), anonymous
```

Comment:
> (indicated by no service name), anonymous clients can ...

This is better expressed in code:

```
if self is XPCAnonymousServer {
    self.connection = nil
}
```

Reply:
Hmm yeah I considered this, but felt it was cleaner to not directly reference a subclass in the superclass and instead have it act based on this server knowing it doesn't correspond to a service.

Reply:
To be precise, the fact it's anonymous isn't actually the relevant bit. (However, only anonymous clients have this behavior.) The relevant aspect is there is no *service*, there's just the listener connection. (As an aside, for this reason `XPCAnonymousServiceClient` is actually misnamed - I'll put up a PR to rename it to just `XPCAnonymousClient`.) Because there's no service, there's no for an interrupted connection be to reestablished.

I'm going to leave this as-is for now. But also agree that neither what I have now nor what you're proposing is entirely "pure" from a class inheritance point of view.

Reply:
The listener connection is anonymous. My interpretation is it's anonymous in the sense the listener connection is not associated with a named service managed by launchd.

Reply:
Discussion continued in #48